content/source/docs/enterprise/sentinel/import/tfconfig.html.md

```diff
@@ -269,27 +269,32 @@ main = rule { tfconfig.module_paths contains ["foo"] }
 
 #### Iterating through modules
 
-Iterating through all modules to find particular resources can be useful. This
-example shows how to use `module_paths` with the [`module()`
-function](#root-function-module) to retrieve all resources of a particular type from
-all modules (in this case, the [`azurerm_virtual_machine`][ref-tf-azurerm-vm]
-resource). Note the use of `else []` in case some modules don't have any
-resources; this is necessary to avoid the function returning undefined.
-
-[ref-tf-azurerm-vm]: /docs/providers/azurerm/r/virtual_machine.html
+Iterating through all modules to find particular resources can be useful. The
+following example function shows how to use `module_paths` with the
+[`module()`](#root-function-module) function to retrieve all resources of a
+particular type from all modules.
 
 ```python
-import "tfconfig"
-
-get_vms = func() {
-	vms = []
-	for tfconfig.module_paths as path {
-		vms += values(tfconfig.module(path).resources.azurerm_virtual_machine) else []
-	}
-	return vms
+find_resources_from_config = func(type) {
+  resource_maps = {}
+  for tfconfig.module_paths else []  as path {
+    resource_maps[path] = tfconfig.module(path).resources[type] else {}
```

Comment:
Recommended to use `strings.join` on the path here so that the key is a string and not a list.

Reply:
The facts that a module path like `path` is a list and that I do want to eventually print out the complete address of each resource or resource instance that violates a policy is why I originally included the second function for each page, either `get_instance_address` for tfstate and tfplan or `get_address` for tfconfig.

Note that while https://docs.hashicorp.com/sentinel/language/spec#map-literals does say that keys of maps should only be boolean, numeric, or string, my function has been working correctly.

Building the correct path is not as simple as using a single join since we need to treat the root module differently than other modules. We could build the module portion of resource addresses inside the find_resources_from_plan function like this, which I just tested:
```
find_resources_from_plan = func(type) {
  resource_maps = {}
  for tfplan.module_paths as path {
    if length(path) == 0 {
      joined_path = ""
    } else {
      joined_path = "module." + strings.join(path, ".module.") + "."
    }
    resource_maps[joined_path] = tfplan.module(path).resources[type] else {}
  }
```

This lets me write my get_instance_address function more simply than before as:
```
get_instance_address = func(module_path, type, name, index) {
  address = module_path + type + "." + name + "[" + string(index) + "]"
  return address
}
```
Note that I included the final "." at the end of joined_path in the first function so I would not need to use an if/else condition in the second function.  My functions adhere to the specification given in https://www.terraform.io/docs/internals/resource-addressing.html since this is what customers are used to seeing in plan and apply logs.  Were you suggesting that this is changing in Terraform 0.12? I don't understand your inclusion of `[data.]` in your suggested format.

If you prefer the above find_resources_from_plan over my original one, I will make the change and ensure it is correctly formatted.  Of course, this would all be easier if Sentinel just gave the joined path itself as you suggested it might in a future helper function.  I would hope that function would give resource addresses the same way that plan and apply logs do.